File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/modules/i18n/TranslationFile.kt
package ca.logaritm.dezel.modules.i18n

import android.content.Context
import ca.logaritm.dezel.extension.getFile
import java.io.InputStream
import java.util.*

/**
 * Translation file locator and loader.
 * @class TranslationFile
 * @since 0.5.0
 */
open class TranslationFile(context: Context, name: String? = null) {

	//--------------------------------------------------------------------------
	//  Properties
	//--------------------------------------------------------------------------

	/**
	 * Whether the translation file has been loaded.
	 * @property loaded
	 * @since 0.5.0
	 */
	public var loaded: Boolean = false
		private set

	/**
	 * @property context
	 * @since 0.5.0
	 * @hidden
	 */
	private var context: Context = context

	//--------------------------------------------------------------------------
	//  Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.5.0
	 */
	init {
		this.load(name)
	}

	//--------------------------------------------------------------------------
	//  Private API
	//--------------------------------------------------------------------------

	/**
	 * @method load
	 * @since 0.5.0
	 * @hidden
	 */
	private fun load(name: String?) {

		if (name == null) {

			val lan = Locale.getDefault().language
			val cou = Locale.getDefault().country

			if (lan == "" ||
				cou == "") {
				return
			}

			this.getFile("${lan}_${cou}.mo")?.let {
				this.load(it)
				return
			}

			this.getFile(Regex("^${lan}_[a-zA-Z]+\\.mo$"))?.let {
				this.load(it)
				return
			}

			return
		}

		this.getFile(name)?.let {
			this.load(it)
		}
	}

	/**
	 * @method load
	 * @since 0.5.0
	 * @hidden
	 */
	private fun load(file: InputStream) {
		TranslationManagerExternal.load(file.readBytes())
		this.loaded = true
	}

	/**
	 * @method getFile
	 * @since 0.5.0
	 * @hidden
	 */
	private fun getFile(named: String): InputStream? {
		return this.context.assets.getFile("languages", named)
	}

	/**
	 * @method getFile
	 * @since 0.5.0
	 * @hidden
	 */
	private fun getFile(match: Regex): InputStream? {
		return this.context.assets.getFile("languages", match)
	}
}

File: platforms/ios/Dezel/Dezel/Modules/I18n/TranslationFile.swift
/**
 * Translation file locator and loader.
 * @class TranslationFile
 * @since 0.5.0
 */
open class TranslationFile {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * Whether the translation file has been loaded.
	 * @property loaded
	 * @since 0.5.0
	 */
	private(set) public var loaded: Bool = false

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * Finds a translation file for the specified language and country.
	 * @constructor
	 * @since 0.5.0
	 */
	public init() {

		let lan = Locale.current.language
		let reg = Locale.current.region

		if (lan == "" ||
			reg == "") {
			return
		}

		if let path = self.getFile(named: "\(lan)_\(reg).mo") {
			self.load(path)
			return
		}

		if let path = self.getFile(match: "^\(lan)_[a-zA-Z]+\\.mo$") {
			self.load(path)
			return
		}
	}

	/**
	 * Finds a translation file with the specified path.
	 * @constructor
	 * @since 0.5.0
	 */
	public init(file: String) {
		if let path = self.getFile(named: file) {
			self.load(path)
		}
	}

	//--------------------------------------------------------------------------
	// MARK: Private API
	//--------------------------------------------------------------------------

	/**
     * @method load
     * @since 0.5.0
     * @hidden
     */
	private func load(_ path: String) {

		do {

			var data = try Data(contentsOf: URL(fileURLWithPath: path))

			data.withUnsafeMutableBytes {
				DLTranslationManagerLoad($0)
			}

			self.loaded = true

		} catch {
			NSLog("Unable to load translation file \(path)")
		}
	}

	/**
	 * @method getFile
	 * @since 0.5.0
	 * @hidden
	 */
	private func getFile(named file: String) -> String? {
		return Bundle.main.getFile(in: "app/languages", named: file)
	}

	/**
	 * @method getFile
	 * @since 0.5.0
	 * @hidden
	 */
	private func getFile(match regex: String) -> String? {
		return Bundle.main.getFile(in: "app/languages", match: regex)
	}
}
